src/workflows/ingestion/conversion/main.nf

```diff
@@ -1,108 +1,42 @@
-            | filter{ it[1].input_type == "10xh5" }
-            | from_10xh5_to_h5mu.run(commonOptions)
-
-        ch_10xmtx = preprocessed_ch
-            | filter{ it[1].input_type  == "10xmtx" }
-            | from_10xmtx_to_h5mu.run(commonOptions)
-
-        ch_h5ad = preprocessed_ch
-            | filter{ it[1].input_type  == "h5ad" }
-            | from_h5ad_to_h5mu.run(commonOptions)
-
-        
-        // /* Combine the different conversion channels */
-        all_ch = ch_10xh5 
-            | mix( ch_10xmtx, ch_h5ad )
-        output_ch = all_ch
-    emit:
-        output_ch
-}
-
-workflow test_wf {
-  helpMessage(config)
-
-  // allow changing the resources_test dir
-  params.resources_test = params.rootDir + "/resources_test"
-
-  // or when running from s3: params.resources_test = "s3://openpipelines-data/"
-
-  testParams = [
-    param_list: [
-      [
-        id: "10xh5_test",
-        input: params.resources_test + "/pbmc_1k_protein_v3/pbmc_1k_protein_v3_filtered_feature_bc_matrix.h5",
-        input_type: "10xh5",
-        modality: null
-      ],
-      [
-        id: "10xmtx_test",
-        input: params.resources_test + "/pbmc_1k_protein_v3/pbmc_1k_protein_v3_filtered_feature_bc_matrix",
-        input_type: "10xmtx",
-        modality: null,
-        output: "\$id.h5mu"
-      ],
-      [
-        id: "10xmtx",
-        input: params.resources_test + "/pbmc_1k_protein_v3/pbmc_1k_protein_v3_filtered_feature_bc_matrix",
-        input_type: "10xmtx",
-        modality: "rna",
-        output: "\$key.h5mu"
-      ],
-      [
-        id: "h5ad",
-        input: params.resources_test + "/pbmc_1k_protein_v3/pbmc_1k_protein_v3_filtered_feature_bc_matrix_rna.h5ad",
-        input_type: "h5ad",
-        modality: "rna",
-        output: "\$key.h5mu"
+  take:
+  input_ch
+
+  main:
+  output_ch = input_ch
+    // translate the input_type argument to the component name that needs to be run
+    | map { id, state ->
+      def componentNameMapper = [
+        "10xh5": from_10xh5_to_h5mu,
+        "10xmtx": from_10xmtx_to_h5mu,
+        "h5ad": from_h5ad_to_h5mu
       ]
-    ]
-  ]
-    
-  output_ch =
-    channelFromParams(testParams, config)
-    | view { "Input: $it" }    
-    | run_wf
-    | view { output ->
-        assert output.size() == 2 : "outputs should contain two elements; [id, file]"
-        assert output[1].output.toString().endsWith(".h5mu") : "Output file should be a h5mu file. Found: ${output[1]}"
-        "Output: $output"
-      }
-      | toSortedList()
-      | map { output_list ->
-        assert output_list.size() == 4 : "output channel should contain four events"
-      }
+      def component = componentNameMapper[state.input_type]
+      def new_state = state + ["component": component]
+      [id, new_state]
+    }
+    | runEach(
+      components: [from_10xh5_to_h5mu, from_h5ad_to_h5mu, from_10xmtx_to_h5mu],
+      filter: { id, state, component ->
+        state.component == component
```

Comment:
I'd avoid storing the component inside the state.

```suggestion
        def componentNameMapper = [
          "10xh5": "from_10xh5_to_h5mu",
          "10xmtx": "from_10xmtx_to_h5mu",
          "h5ad": "from_h5ad_to_h5mu"
        ]
        componentNameMapper[state.input_type] == component.config.functionality.name
```

Reply:
done in 5bd7217c6c0dc779b8f89a52e3fd196e087aa0e8